### ISSUE TITLE
`DistBag_DFS`: our revisited version of `DistBag` for depth-first tree-search

### ISSUE THREAD

**Issue by Guillaume-Helbecque** [chapel-lang/chapel#21958]
This issue presents the first version of our `DistBag_DFS` distributed data structure, which is a revisited version of the Chapel's `DistBag` data structure, supplied in the `DistributedBag` package module. This work is in progress, and this issue is intended to discuss its development as well as possible integration into the language.

### Background
Our research focuses on productivity-aware parallel distributed Depth-First tree-Search (DFS) algorithms. On the one hand, DFS is frequently used in combinatorial algorithms because memory requirements of other search strategies often become excessive, and is easily implemented by storing generated, but not yet visited nodes in a stack (LIFO). On the other hand, we focus on the parallel tree exploration model, which consists in exploring several disjoint subspaces in parallel, meaning that multiple DFS, rooted in different tree nodes, are performed in parallel. In asynchronous mode, adopted in our work, the search processes communicate in an unpredictable manner and the sharing of knowledge among workers becomes non-trivial. Therefore, defining a data structure to store the work pool and an associated management policy is highly important for performance.

### Limitations of `DistBag`
In this context, `DistBag` was particularly attractive to us. Indeed, it provides a multi-set implementation (called *segments*) hidden to the user, as well as an underlying load-balancing mechanism, which is necessary to solve irregular problems. However, preliminary experiments and source-code inspection revealed that its segments' (pools') scheduling policy does not allow us to use it for parallel DFS, nor to have any control over the order of insertion/retrieval of nodes. Indeed, pools do not necessarily operate in LIFO-order and are not explicitly mapped onto threads. In fact, multiple pools are maintained, but threads remove and insert elements from any (not necessarily the same) unlocked pool. Although this may be acceptable for some applications, this behaviour is not suitable for parallel DFS. As a direct consequence, memory requirements may rapidly grow out of control. This observation is consistent with the documentation which states that important memory consumption may appear in single-locale experiments:

https://github.com/chapel-lang/chapel/blob/8941286fffe7d89561f0b8fd75bd31889973c643/modules/packages/DistributedBag.chpl#L97

In addition, there are already two reported issues on `DistBag`: #19860, #19867. This motivates our revisit.

### What's new in `DistBag_DFS` ?

1. We extended the segments' scheduling policy to support insertion and retrieval from both ends, effectively supporting both FIFO and LIFO orders, like a deque (double-ended queue). At the block level, we added the `popHead` procedure and renamed `pop` and `push` to `popTail` and `pushTail`, respectively. Moreover, we mapped explicitly segments to threads, in order to ensure DFS. This is done by specifying the calling task's id in the insertion/retrieval operations (*e.g* `bag.add(elts, taskId)`, `bag.remove(taskId)`). In this way, we can explicitly control the order of exploration, and ensure that child nodes are inserted into the segment where the parent node was retrieved.

2. We experimented a new synchronization mechanism using non-blocking split deques. In `DistBag`, segments are lock-protected using atomic variables and hand-coded synchronization mechanisms. Instead, we propose to use non-blocking split deques, meaning that each deque is split into a *public* and a *private* portion using an atomic *split pointer*. Under this scheme, all processes push new tasks on the tail of the queue and pop tasks from the tail to get the next task to execute in LIFO order, while work stealing (WS) is done at the head in a FIFO manner. This synchronization scheme allows lock-free local access to the private portion of the deque and copy-free transfer of work between the public and private portions of the deque. Work transfer is done by moving the split pointer in either directions. At the segment level, we re-implemented most of the procedures.

3. Finally, we redefined the underlying WS mechanism according to the last changes. In few words, the latter is a bi-level mechanism, where thief first try to steal locally, and if not possible, globally. We implemented a `victim` iterator that chooses the victim thread/locale according to the specified selection policy (random, ring). Moreover, since global WS operations generate high parallel overheads, the thief is expected to steal more work items than it needs. Then, these extra work items provide work for potential future local WS. Thus, when a thread performs a global steal, other global steal requests issued by threads in its locale are ignored.

The implementation of the `DistBag_DFS` data structure is available in our public repository, and is supplied in the `DistributedBag_DFS` module: https://github.com/Guillaume-Helbecque/P3D-DFS.

### Tutorial
The following code fully enumerates a complete binary tree of a given depth, and determines the size of the tree. It shows the basic use of `DistBag_DFS`, and is not optimised for performance.
```chapel
use List;
use DistributedBag_DFS;

config const N: int = 5;

record Node { var depth: int = 0; }

// decompose a parent node into children ones
proc decompose(const parent: Node, treeSize: atomic int)
{
  var children: list(Node);

  if (parent.depth < N) {
    children.append(new Node(parent.depth + 1)); // left child
    children.append(new Node(parent.depth + 1)); // right child
    treeSize.add(2); // increase counter
  }

  return children;
}

proc main()
{
  var treeSize: atomic int;

  var bag = new DistBag_DFS(Node, Locales); // initialization
  var root = new Node();

  bag.add(root, 0); // insert the root node in task-0's segment (arbitrarily)

  coforall loc in Locales do on loc {
    coforall taskId in 0..#here.maxTaskPar {
      while true {
        var (hasWork, parent) = bag.remove(taskId); // remove a parent node from taskId's segment

        if (hasWork == 1) {
          bag.addBulk(decompose(parent, treeSize), taskId); // insert children nodes in the taskId's segment
        }
        else {
          break;
        }
      }
    }
  }

  writeln("Tree depth: ", N+1, " and tree size: ", treeSize.read()+1);
  return 0;
}
```

### Experimental results
We experimented `DistBag_DFS` on large unbalanced tree-based problems, which are the Permutation Flowshop Scheduling Problem (PFSP), and the Unbalanced Tree-Search (UTS) benchmark. Our algorithm is named P3D-DFS and publicly available on Github: https://github.com/Guillaume-Helbecque/P3D-DFS. 
The experiments have been conducted on a cluster equipped with 2 AMD Epyc ROME 7H12 @ 2.6 GHz processors, including a total of 128 cores and 256 GB RAM per computer nodes. All nodes are interconnected through a Fast InfiniBand HDR100 network, and operate under Red Hat Enterprise Linux 8.3, 64 bits. The implementation is based on Chapel 1.29.0.
P3D-DFS has been compared to state-of-the-art baseline implementations in terms of performance, using up to 64 computer nodes including each 128 cores each. In a shared-memory setting, the experimental results revealed that P3D-DFS is competitive with its OpenMP counterparts for coarser-grained instances, while it suffers from parallel overheads at low granularity. Moreover, by using `DistBag_DFS`, one can achieve performance equivalent to MPI+X in a distributed-memory setting, but with a programming effort much smaller. As an example, please find below the comparison between P3D-DFS and a MPI+pthread baseline (called MPI-PBB) at the distributed-memory level, on the Branch-and-Bound (B&B) method applied to the PFSP.

<p align="center">
<img src="https://user-images.githubusercontent.com/72358009/227473495-0dd4ab0b-d5a1-450d-89f9-1184327a8d1b.png" width="75%" height="75%">
</p>

### Future works & open questions
- Improve the internal mechanisms to allow better performance for fine-grained applications.
- Although the "split deque" synchronization mechanism provides scalable performance until 64 nodes (8192 cores), its implementation and extension are difficult. Indeed, in `DistBag` segments are implemented as "unrolled linked-blocks", and therefore have a dynamic size. In `DistBag_DFS`, segments are only made up of a single block of fixed size, and we question the challenge and need of implementing a "split unrolled linked block" for example.
- Should `DistBag` and `DistBag_DFS` be wrapped into a single data structure (let's say `DistBag`), where the DFS policy is an optional parameter? Something like:
```chapel
/* We could imagine a third optional initialization parameter for DistBag (let say 'explorationPolicy'),
which could be set to 'DFS' in order to internally use 'DistBag_DFS', or 'none' to use 'DistBag' by default.*/
var new bag = DistBag(int, Locales, explorationPolicy = DFS);
```
In that case, the calling task's id become also an optional parameter of the insertion and retrieval procedures.

**Comment by mppf** [comment]
hi @Guillaume-Helbecque -- thanks for your efforts in this area and writing it up in this issue. I think would be really great to see your efforts included somehow as improvements to the DistBag module that Chapel includes in modules/packages.

I have two questions for you:

> Should `DistBag` and `DistBag_DFS` be wrapped into a single data structure (let's say `DistBag`), where the DFS policy is an optional parameter? Something like:

One thing about this is unclear to me. Suppose we entirely replaced the package module `DistBag` with your `DistBag_DFS` effort. What would we be losing? While I know there were some performance studies with `DistBag` at the time it was created, I don't think these were as motivated by applications as your work. So, if the only concern is that there could be applications which get slower, I'm not immediately worried about that (but, of course, we could ask on the chapel-users mailing list, say, to see if anybody using DistBag has concerns with this approach). Of course you might know of things we would be losing that I am not aware of.


> 1. We extended the segments' scheduling policy to support insertion and retrieval from both ends, effectively supporting both FIFO and LIFO orders, like a deque (double-ended queue).

I was tangentially involved in the DistBag effort when it was originally developed and at the same time, [DistDeque](https://chapel-lang.org/docs/modules/packages/DistributedDeque.html) was added. It sounds like your extension here gets DistBag a bit closer to DistDeque but I wanted to ask if you would consider `DistBag_DFS` to be a deque or if it just has some deque-like properties that help it to optimize locally. (Vs `DistDeque` which offers a global total order.)

**Comment by Guillaume-Helbecque** [comment]
Hi @mppf, thank you for your interest on `DistBag-DFS`.

> Suppose we entirely replaced the package module DistBag with your DistBag_DFS effort. What would we be losing?

First of all, it would be interesting to know for which type of applications `DistBag` was originally developed. Indeed, it is true that some micro-benchmarks have been investigated when it was created, nevertheless, to the best of our knowledge, there is no applicative study using the data structure. As you proposed, it would be interesting to ask on the chapel-users mailing list if anybody uses `DistBag`, and for what purpose.

That being said, `DistBag-DFS` has been specifically developed for DFS, and we question its use/extension to other search strategies (*e.g* breadth-first, best-first). Indeed, the new synchronization mechanism based on non-blocking split deque is particularly suitable for DFS since elements are inserted/retrieved in a LIFO manner, *i.e.* from the same entry of the deque. However, in its current state, `DistBag-DFS` doesn't allow FIFO for example, since it implies insertion/retrieval operations from both ends of the deque (not possible because one is in the private part while the other in the public one). May be `DistBag` is more general in that point.

From our point of view, `DistBag-DFS` is more a special case of `DistBag` than a replacement. 

> I wanted to ask if you would consider DistBag_DFS to be a deque or if it just has some deque-like properties that help it to optimize locally.

In fact, `DistBag-DFS` can be viewed as a collection of deques, rather than a deque itself. The reason is that there is no global ordering of the elements (no global DFS). We only guarantee DFS locally to each task. And as you said, the deque-like properties help us to optimize locally since it allows the non-blocking split deque synchronization scheme.

**Comment by Guillaume-Helbecque** [comment]
Following the presentation of `DistBag-DFS` at CHIUW 2023, we would like to enhance the data structure (DS) a bit more. we are currently investigating possible improvements in terms of performance, extending our experiments to large scale cluster (~50k cores), and planning more dissemination.

Since the DS is expected to serve the Chapel community, we would be pleased to have your expert point of view regarding the following points:
- At this stage of development, is it possible to integrate the DS in Chapel as a package module? For example, as a sibling of `DistributedBag`. This would considerably increase its visibility, and potentially enable us to receive user comments that would help us to improve its development and its relationship with `DistBag`.
- We already validated the DS on real-world large unbalanced tree-based problems (see above: PFSP), and we now plan to implement some micro-benchmarks in order to test the low-level mechanisms and to compare with existing DS, as it has been done for `DistBag` (https://chapel-lang.org/releaseNotes/1.16/03-library.pdf, slides 13-17). Except classical insertion/retrieval operations, are there any other test that we could perform and that could be especially significant for you?
- Along with the DS, we could also contribute by providing some nightly testing to ensure that the DS kept working as the language and compiler continued to evolve. One test that could be of interest is the Unbalanced tree-search benchmark, which was already tested using `DistBag-DFS` (see CHIUW 2023: https://chapel-lang.org/CHIUW/2023/Helbecque.pdf). Our current support for UTS rely on external C code, but I guess you would prefer a full-Chapel version?

**Comment by mppf** [comment]
> At this stage of development, is it possible to integrate the DS in Chapel as a package module? 

I think so, but we might need some other people to be on board with it as well. Note that, eventually we'll want to move all these package modules to `mason` packages or to standard library modules; but that's a bit of a project on its own and I'm not sure when we will get to it. (For one thing, we'll have get back to `mason` development which is currently waiting behind other efforts)

I think it's likely that we'll want to adjust the name DistBag_DFS, also. Ideally, any user-facing parts of a new package module should follow the [standard library style guide](https://github.com/chapel-lang/chapel/blob/main/doc/rst/developer/bestPractices/StandardModuleStyle.rst) -- in particular, we won't want to see the underscore in the name.

> Except classical insertion/retrieval operations, are there any other test that we could perform and that could be especially significant for you?

This might be a good question for the [Chapel Users discourse](https://chapel.discourse.group/c/users).

For my part, I'm interested in seeing some distributed data structures that can be used as a work queue, but that might or might not be a reasonable use of DistBag_DFS.

> Along with the DS, we could also contribute by providing some nightly testing to ensure that the DS kept working as the language and compiler continued to evolve.

That would be great. If you make a PR adding such a package module, we'll need tests for it. Some amount of portability for such tests is important but I don't think it's a big problem if it relies on some C code. I'd be more concerned about the tests being really complicated, if that's what a version using your existing UTS would do.

See also https://chapel-lang.org/docs/developer/bestPractices/ContributorInfo.html if you have not already.

**Comment by bradcray** [comment]
> Ideally, any user-facing parts of a new package module should follow the [standard library style guide](https://github.com/chapel-lang/chapel/blob/main/doc/rst/developer/bestPractices/StandardModuleStyle.rst) -- in particular, we won't want to see the underscore in the name.

I don't think of it as being a requirement that package modules follow the standard library style guide (since they're package, rather than standard, modules).  That said, doing so has the benefits of them seeming more similar to the standard libraries and more ready for being converted into a standard module if/when we wanted to do this.  But I wouldn't (personally) let that factor slow down progress on merging the PR so that others can start using it...

I have a question about whether we should just have this DistBag replace the previous rather than sitting alongside it.  @Guillaume-Helbecque and @mppf , do either of you have opinions on that?  I think my motivation is that the previous isn't being actively maintained, is more likely to be something that people reach for (given the simpler / more straightforward name), and has had issues raised against it over time.  Alternatively, we could keep both, but use the simpler names for Guillaume's effort since it's receiving more attention and care.  If we only kept the new one, it'd be interesting to see the level of effort to convert the old tests over to the new interface.  If we kept both, and any of the old tests had value, it'd be nice to see clones of them that worked with the new implementation.

Guillaume, you may know, but in addition to correctness tests, we also [run nightly performance tests](https://chapel-lang.org/perf-nightly.html), so that could be another way to track the performance of the package over time (Where I think a UTS benchmark would be nice if it's tractable to get it into our testing system).  I wouldn't worry about performance tests in the first PR, though, and would handle those in subsequent PRs.

**Comment by mppf** [comment]
> I have a question about whether we should just have this DistBag replace the previous rather than sitting alongside it. 

In case you missed it, I asked about this a [bit earlier in this issue](https://github.com/chapel-lang/chapel/issues/21958#issuecomment-1521749209) (in case you missed that):

> From our point of view, `DistBag-DFS` is more a special case of `DistBag` than a replacement. 


I also lean towards replacing it. Using the simpler name for the new effort could work as well, if we are nervous about removing functionality. Re. removing the old implementation, it might be something to ask on the Chapel Users discourse -- if anybody is using DistBag in a way that wouldn't be improved by the new DistBag-DFS. Of course, it will be easier for users of DistBag to answer such questions if they can simply try it, which argues for keeping the old one with a different name.

**Comment by mppf** [comment]
> > This might be a good question for the Chapel Users discourse.
> 
> May I ask you to do this for me? I'm not very familiar with this tool.

I posted it in https://chapel.discourse.group/t/are-you-using-distbag/24628

**Comment by Guillaume-Helbecque** [comment]
I just created a draft PR to add the data structure as well as a set of correctness tests. While the module and its doc are almost ready, I marked this PR as a draft because several points remain to be discussed, e.g., the name of the data structure and its collaboration with `DistBag`.

@mppf - As far as I know, there has been no reply to your message posted on the Chapel Users discourse, is that right? The holiday period has probably disconnected people a bit, but I sense that this is a first indicator of the use of `DistBag`.

**Comment by mppf** [comment]
> @mppf - As far as I know, there has been no reply to your message posted on the Chapel Users discourse, is that right? The holiday period has probably disconnected people a bit, but I sense that this is a first indicator of the use of `DistBag`.

Yes, no replies, and I haven't received any reply to it off-list either.

**Comment by Guillaume-Helbecque** [comment]
@bradcray, @mppf : One roadblock towards PR #23204 is the Chapel-standardization of the name of the data structure, and the coexistence of both `DistBag` and `DistBag_DFS` package modules. These questions are out of my reach, and I wonder if your thinking has evolved/matured since we last spoke.

**Comment by bradcray** [comment]
@Guillaume-Helbecque : I'm very sorry for the slow response here.  This arrived during a busy period and I'm just finding my way back to it now.

I continue to worry that the default `DistributedBag` has not gotten enough attention or love to warrant owning the obvious/main name, and would have no qualms about having it be renamed and Guillaume's effort taking over the main name.  That said, at this point, I'm more interested in making forward progress than worrying overmuch about the names, so am not at all insistent about this if it's become the barrier to progress.

@mppf: What would you like to do here to move this forward?

**Comment by mppf** [comment]
@bradcray - I think we are in agreement about the names and also not wanting it to hold things up. In fact, I would lean towards replacing the existing `DistributedBag` module with the new effort, now that we have asked if anybody else is using it (and got no replies). If we wish to preserve both, we have to choose names for both. However, we are talking about package modules here, so they are not stable & we can change the names and/or functionality in the future.

Regarding names within the module, it looks to me like the main name we are uncertain about is `record DistBag_DFS`. If we are going with replacing DistributedBag with the new effort, this would ideally be named `record distBag`. (Even if we have 2 modules, we can give them the same type name).

@Guillaume-Helbecque - are there other names you are worried about? Do you feel OK about replacing the existing `DistributedBag` now that we asked if anybody was using it? I understand that your effort is optimized for DFS, but in my opinion that can be reflected in the documentation and does not have to be recorded in the name.

**Comment by Guillaume-Helbecque** [comment]
> However, we are talking about package modules here, so they are not stable & we can change the names and/or functionality in the future.

You are right to mention that the current version of `DistBag_DFS` is unstable, and may change in the future.

> Are there other names you are worried about? 

If we choose to change the name to `DistBag`, we will need to emphasise the DFS aspect in the documentation, as you mention, and may be quote the old name `DistBag_DFS`, in order to make easier the link with my research work, introducing the data structure under that old name.

>Do you feel OK about replacing the existing `DistributedBag` now that we asked if anybody was using it? 

I assume that this is the best option, yes. Nevertheless, I think we should keep both versions in the first instance, may be deprecate the existing one, and see if there is any feedback from users regarding the new one.

**Comment by mppf** [comment]
Thanks @Guillaume-Helbecque - that all sounds good to me. If we are keeping the old DistributedBag around with a different name and deprecating it, then we have to choose a name for it. I would propose simply naming the module / file `DistributedBagDeprecated` and then we can remove that after a release if nobody needs it.

So, summing up:
 * `modules/packages/DistributedBag.chpl` contains your new DFS-optimized work & probably should rename `record DistBag_DFS` to `record distBag`.
 * `modules/packages/DistributedBagDeprecated.chpl` contains the old version, but is deprecated and will be removed

Are there any remaining open questions that are holding up the PR?
